transaction/ResourceId.hpp

```diff
@@ -0,0 +1,217 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_TRANSACTION_RESOURCE_ID_HPP_
+#define QUICKSTEP_TRANSACTION_RESOURCE_ID_HPP_
+
+#include <limits>
+#include <string>
+#include <type_traits>
+
+#include "catalog/CatalogTypedefs.hpp"
+#include "storage/StorageBlockInfo.hpp"
+#include "utility/HashPair.hpp"
+#include "utility/Macros.hpp"
+
+namespace quickstep {
+
+namespace transaction {
+
+/** \addtogroup Transaction
+ *  @{
+ */
+
+/**
+ * @brief Class for representing unique id for each database resource
+ *        e.g Tuple, block, relation, database etc...
+ **/
+class ResourceId {
+ public:
+  static constexpr database_id kDbIdPlaceholder =
+    std::numeric_limits<database_id>::max();
+  static constexpr relation_id kRelIdPlaceholder =
+    std::numeric_limits<relation_id>::max();
+  static constexpr block_id kBlockIdPlaceholder =
+    std::numeric_limits<block_id>::max();
+  static constexpr tuple_id kTupleIdPlaceholder =
+    std::numeric_limits<tuple_id>::max();
```

Comment:
The main thing here is about consistency so that the values we choose here should be invalid. If we choose to use the max values, we need to change the typedefs for `database_id` and `relation_id` etc, and mark the max values as invalid.